src/lib/SaferNames/Syntax.hs

```diff
@@ -436,147 +349,392 @@ instance Semigroup (SourceNameMap n) where
-instance SubstE EffectRow where
-  traverseCoreE = undefined
-
-instance SubstE e => HasNamesE (SubstVal e) where traverseNamesE = traverseNamesFromSubstE
-instance SubstE e => SubstE (SubstVal e) where
-  traverseCoreE = undefined
-
-instance SubstE e => HasNamesE (WithArrow e) where
-  traverseNamesE = traverseNamesFromSubstE
-instance SubstE e => SubstE (WithArrow e) where
-  traverseCoreE s t (WithArrow arr e) =
-    WithArrow <$> traverse (tce s t) arr <*> tce s t e
-
-instance HasNamesE Block where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Block where
-  traverseCoreE s t (Block decls result) =
-    traverseCoreE s t (Abs decls result) >>= \case
-      Abs decls' result' -> return $ Block decls' result'
+(<++>) :: String -> String -> String
+(<++>) s1 s2 = s1 <> " " <> s2
 
-instance HasNamesB Decl where
-  traverseNamesB = traverseNamesFromSubstB
-  boundScope = undefined
-instance SubstB Decl where
-  traverseCoreB s t (Let ann b expr) = do
-    expr' <- traverseCoreE s t expr
-    traverseCoreB s t b >>= \case
-      FreshBinder ext b' renamer ->
-        return $ FreshBinder ext (Let ann b' expr') renamer
-  asBindingsFrag = undefined
-
-instance HasNamesB DataConRefBinding where
-  traverseNamesB = undefined
-  boundScope = undefined
-
-instance SubstB b => SubstB (Nest b) where
-  traverseCoreB s t nest = case nest of
-    Empty -> do return $ FreshBinder emptyFreshExt Empty emptyEnv
-    Nest b rest ->
-      traverseCoreB s t (NestPair b rest) >>= \case
-        FreshBinder ext (NestPair b' rest') renamer ->
-          return $ FreshBinder ext (Nest b' rest') renamer
-  asBindingsFrag = undefined
-
-instance (SubstB b1, SubstB b2) => SubstB (NestPair b1 b2) where
-  traverseCoreB = undefined
-  asBindingsFrag = undefined
-
-instance HasNamesE NamedDataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE NamedDataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE DataDef where traverseNamesE = traverseNamesFromSubstE
-instance SubstE DataDef where
-  traverseCoreE = undefined
-
-instance HasNamesE Atom where traverseNamesE = traverseNamesFromSubstE
-instance SubstE Atom where
-  traverseCoreE s t atom = case atom of
-    Var (AnnVar v ann) ->
-      lookupNameTraversal t v >>= \case
-        Left substVal -> case substVal of
-          SubstVal x -> return x
-          Rename v' -> do
-            ann' <- traverseCoreE s t ann
-            return $ Var $ AnnVar v' ann'
-        Right v' -> do
-          ann' <- traverseCoreE s t ann
-          return $ Var $ AnnVar v' ann'
-    Lam ab -> Lam <$> tce s t ab
-    Pi  ab -> Pi  <$> tce s t ab
+showParens :: Show a => a -> String
+showParens x = "(" <> show x <> ")"
+
+instance Show (DataDef n) where
+  show (DataDef name bs cons) =
+    "DataDef" <++> showParens name <++> showParens bs <++> showParens cons
+
+instance SubstE AtomSubstVal (Name DataDef) where
+  substE name =
+    lookupSubstM name >>= \case
```

Comment:
Style question: I probably would have written
```
do
  Rename name' <- lookupSubstM name
  return name'
```
here. Is there a reason to prefer explicit bind with a lambda case?

Reply:
Good question. Definitely the `do` version is much more readable. The reason I started using lambda case when unpacking GADTs is that it seemed to work better at bringing the GADT's type classes and existential types into scope. But more recently I've noticed that `<-` sometimes does too. So maybe it's just `let`-within-`do` that doesn't? I should probably just learn what the rules are so I can make sure I only use the lambda case style when it's absolutely necessary.